shared/extension/Behaviour.hpp

```diff
@@ -0,0 +1,317 @@
+            /// The action type that this provider is using (normal, entering, leaving)
+            ProviderAction action;
+        };
+
+        /**
+         * Message used to tell the Director about a new When condition on a provider
+         */
+        struct WhenExpression {
+            WhenExpression(const std::shared_ptr<NUClear::threading::Reaction>& reaction,
+                           const std::type_index& type,
+                           bool (*validator)(const int&),
+                           int (*current)())
+                : reaction(reaction), type(type), validator(validator), current(current) {}
+
+            /// The provider reaction that is contingent on this when condition
+            std::shared_ptr<NUClear::threading::Reaction> reaction;
+            /// The enum type that this when expression is looking at
+            std::type_index type;
+            /// Function to determine if the passed state is valid
+            bool (*validator)(const int&);
+            /// Function to get the current state from the global cache
+            int (*current)();
+            /// Function to bind a reaction to monitor when this state changes
+            NUClear::threading::ReactionHandle (*binder)(NUClear::Reactor&, std::function<void(const int&)>);
+        };
+
+        /**
+         * Message used to tell the director about a new Causing relationship for this provider
+         */
+        struct CausingExpression {
+            CausingExpression(const std::shared_ptr<NUClear::threading::Reaction>& reaction,
+                              const std::type_index& type,
+                              const int& resulting_state)
+                : reaction(reaction), type(type), resulting_state(resulting_state) {}
+            /// The provider reaction that will cause this state
+            std::shared_ptr<NUClear::threading::Reaction> reaction;
+            /// The enum type that this when expression claiming to manipulate
+            std::type_index type;
+            /// The resulting state this provider is claiming to provide if it is executed
+            int resulting_state;
+        };
+
+        /**
+         * Message to tell the Director that a provider task has finished executing
+         */
+        struct ProviderDone {
+            ProviderDone(const uint64_t& requester_id, const uint64_t& requester_task_id)
+                : requester_id(requester_id), requester_task_id(requester_task_id) {}
+            /// The reaction_id of the provider that finished
+            uint64_t requester_id;
+            /// The specific task_id of the provider that finished
+            uint64_t requester_task_id;
+        };
+
+        /**
+         * A task message that is to be sent to the Director for execution
+         */
+        struct DirectorTask {
+            DirectorTask(std::type_index type,
+                         uint64_t requester_id,
+                         uint64_t requester_task_id,
```

Comment:
placing parameters of the same type next to each other makes it easier to screw up using your interface

Reply:
True, but I still want both of these data types, and hopefully, the only person calling these internal extension messages in Behaviour will be the DSL words below. Nobody should be making these outside of this file in the emit<Task> extension.